### PR TITLE
util/chunk: add mutable row type.

### PR DESCRIPTION
**Pull request by coocood**
`MutRow` has `Set` methods and can be easily created and converted to chunk.Row, can be used for write statements.

### CODE REVIEW

**Comment by coocood** [comment]
@winkyao @zz-jason PTAL

File: util/chunk/mutrow.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+func MutRowFromDatums(datums []types.Datum) MutRow {
+	c := &Chunk{}
+	for _, d := range datums {
+		col := makeMutRowColumn(d.GetValue())
+		c.columns = append(c.columns, col)
+	}
+	return MutRow{c: c, idx: 0}
+}
+
+// MutRowFromTypes creates a MutRow from a datum slice, each column is initialized to default
```
Comment (XuHuaiyu) [inline util/chunk/mutrow.go]: unfinished comment?
s/ a datum slice/ a FieldType slice

File: util/chunk/mutrow.go [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+func (mr MutRow) SetRow(row Row) {
+	for colIdx, rCol := range row.c.columns {
+		mrCol := mr.c.columns[colIdx]
+		elemLen := len(rCol.elemBuf)
+		if elemLen > 0 {
+			copy(mrCol.data, rCol.data[row.idx*elemLen:(row.idx+1)*elemLen])
+		} else if len(rCol.offsets) > 0 {
+			setMutRowBytes(mrCol, rCol.data[rCol.offsets[row.idx]:rCol.offsets[row.idx+1]])
+		} else {
+			mrCol.ifaces[0] = rCol.ifaces[row.idx]
```
Comment (XuHuaiyu) [inline util/chunk/mutrow.go]: mrCol.ifaces[0] will point at the same address as rCol.ifaces[row.idx] ?

File: util/chunk/mutrow.go [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+func setMutRowBytes(col *column, bin []byte) {
+	if len(col.data) >= len(bin) {
+		col.data = col.data[:len(bin)]
+	} else {
+		buf := make([]byte, len(bin)+1)
+		col.data = buf[:len(bin)]
+		col.nullBitmap = buf[len(bin):]
+	}
+	copy(col.data, bin)
+	col.offsets[1] = int32(len(bin))
```
Comment (XuHuaiyu) [inline util/chunk/mutrow.go]: should we set col.nullBitmap[0] = 1?

File: util/chunk/mutrow.go [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+	if len(col.data) >= dataLen {
+		col.data = col.data[:dataLen]
+	} else {
+		buf := make([]byte, dataLen+1)
+		col.data = buf[:dataLen]
+		col.nullBitmap = buf[dataLen:]
+	}
+	binary.LittleEndian.PutUint64(col.data, val)
+	copy(col.data[8:], name)
+	col.offsets[1] = int32(dataLen)
```
Comment (XuHuaiyu) [inline util/chunk/mutrow.go]: set col.nullBitMap[0] = 1

File: util/chunk/mutrow.go (hunk 2 above)
Comment (coocood) [reply util/chunk/mutrow.go]: Yes, the interface value itself is never modified.

File: util/chunk/mutrow.go (hunk 3 above)
Comment (coocood) [reply util/chunk/mutrow.go]: it's set at line 299

File: util/chunk/mutrow.go (hunk 4 above)
Comment (coocood) [reply util/chunk/mutrow.go]: ditto

File: util/chunk/mutrow.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+		setMutRowNameValue(col, s.Name, s.Value)
+	default:
+		mr.c.columns[colIdx] = makeMutRowColumn(d.GetValue())
+	}
+	col.nullBitmap[0] = 1
+}
+
+func setMutRowBytes(col *column, bin []byte) {
+	if len(col.data) >= len(bin) {
+		col.data = col.data[:len(bin)]
```
Comment (XuHuaiyu) [inline util/chunk/mutrow.go]: we may set like this?
``` go
col.data = col.data[:len(bin+1)]
col.nullBitmap = col.data[len(bin):]
```

File: util/chunk/mutrow.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+		col.nullBitmap = buf[len(bin):]
+	}
+	copy(col.data, bin)
+	col.offsets[1] = int32(len(bin))
+}
+
+func setMutRowNameValue(col *column, name string, val uint64) {
+	dataLen := len(name) + 8
+	if len(col.data) >= dataLen {
+		col.data = col.data[:dataLen]
```
Comment (XuHuaiyu) [inline util/chunk/mutrow.go]: ditto

File: util/chunk/mutrow.go (hunk 5 above)
Comment (coocood) [reply util/chunk/mutrow.go]: Then the data length is not the same as `len(bin)`.

**Pull request merged**